astra-sim/astra-sim
# Changhai trace hwrsrc

## Summary
This add some trace logs in the workload::hardware_resources, which provide finer grid infos that which node is issues, how the hardware resources occupancy changes. Also add flags to disable/enable this log in CMakeLists to avoid generating huge logs for everyone.

## Test Plan
1. edit <astra-sim>/CMakeList.txt, uncomment line 55 `target_compile_definitions(AstraSim PRIVATE -DASTRASIM_TRACE_HARDWARE_RESOURCES)`.
2. Rebuild astrasim: `./build/astra_analytical/build.sh -l && ./build/astra_analytical/build.sh`
3. Run example: `cd run/examples && ./run.sh`
Then check log to get the hwrsrc traces, for example
```
[2024-01-25 16:38:46.447] [workload] [trace] try issue node.id=8174
[2024-01-25 16:38:46.447] [workload::HardwareResource] [trace] Current hw resources: cpu_comp_ops=0, gpu_comp_ops=0, comm_ops=0, comm_ops=0
[2024-01-25 16:38:46.447] [workload] [trace] issue success node.id=8174
[2024-01-25 16:38:46.447] [workload::HardwareResource] [trace] Node.id=8174 occupy cpu_comp_ops
[2024-01-25 16:38:46.447] [workload::HardwareResource] [trace] Node.id=8388 release cpu_comp_ops
[2024-01-25 16:38:46.447] [workload] [trace] try issue node.id=8389
[2024-01-25 16:38:46.447] [workload::HardwareResource] [trace] Current hw resources: cpu_comp_ops=0, gpu_comp_ops=0, comm_ops=0, comm_ops=0
[2024-01-25 16:38:46.447] [workload] [trace] issue success node.id=8389
[2024-01-25 16:38:46.447] [workload::HardwareResource] [trace] Node.id=8389 occupy cpu_comp_ops
[2024-01-25 16:38:46.447] [workload::HardwareResource] [trace] Node.id=8603 release cpu_comp_ops
[2024-01-25 16:38:46.447] [workload] [trace] try issue node.id=8604
[2024-01-25 16:38:46.447] [workload::HardwareResource] [trace] Current hw resources: cpu_comp_ops=0, gpu_comp_ops=0, comm_ops=0, comm_ops=0
```

## Additional Notes
Enable lots trace might have an impact on performance, however, for the example run the difference is not observable.
